### PR TITLE
Logging in to Spotify

### PR DESCRIPTION
**Pull request by TOMalley104**

### CODE REVIEW

File: spotifyMini/AppDelegate.swift [hunk 1]
```diff
@@ -12,12 +12,41 @@ import UIKit
 class AppDelegate: UIResponder, UIApplicationDelegate {
 
     var window: UIWindow?
+    var session: SPTSession?
```
Comment (brightredchilli) [inline spotifyMini/AppDelegate.swift]: session and player not used, best to remove until they are actually used.

File: spotifyMini/AppDelegate.swift [hunk 2]
```diff
@@ -12,12 +12,41 @@ import UIKit
...
+    var player: SPTAudioStreamingController?
 
+    let clientID = "7106e0a1f4c94c6e9e3f1b209e11a3fb"
+    let clientSecret = "357c7f83e10a475887927ad3f389068a"
```
Comment (brightredchilli) [inline spotifyMini/AppDelegate.swift]: client secret not used, remove

File: spotifyMini/AppDelegate.swift [hunk 3]
```diff
@@ -12,12 +12,41 @@ import UIKit
...
-        // Override point for customization after application launch.
+        self.setupSpotifyAuth()
         return true
     }
+    
+    func setupSpotifyAuth() {
+        let auth = SPTAuth.defaultInstance()
+        auth.clientID = self.clientID
+        auth.redirectURL = self.callbackURL
+        auth.sessionUserDefaultsKey = "SpotifySession" // FIXME: make constant
```
Comment (brightredchilli) [inline spotifyMini/AppDelegate.swift]: easy enough to just make a constant now for this  :)

File: spotifyMini/AppDelegate.swift [hunk 4]
```diff
@@ -12,12 +12,41 @@ import UIKit
...
+        auth.requestedScopes = [SPTAuthStreamingScope]
+    }
+    
+    func application(application: UIApplication, openURL url: NSURL, sourceApplication: String?, annotation: AnyObject) -> Bool {
+        if SPTAuth.defaultInstance().canHandleURL(url) {
+            SPTAuth.defaultInstance().handleAuthCallbackWithTriggeredAuthURL(url, callback: { error, session in
+                if let error = error {
+                    // FIXME: throw application error
+                    print("openURL error: \(error.localizedDescription)")
+                    return
```
Comment (brightredchilli) [inline spotifyMini/AppDelegate.swift]: does this compile? 😮

File: spotifyMini/AppDelegate.swift [hunk 5]
```diff
@@ -40,6 +69,8 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
     func applicationWillTerminate(application: UIApplication) {
         // Called when the application is about to terminate. Save data if appropriate. See also applicationDidEnterBackground:.
     }
+
```
Comment (brightredchilli) [inline spotifyMini/AppDelegate.swift]: extra lines, some people at intrepid are big sticklers for this...im not so much, but just letting you know : )

File: spotifyMini/AppDelegate.swift [hunk 6]
```diff
@@ -12,12 +12,41 @@ import UIKit
...
+    
+    func application(application: UIApplication, openURL url: NSURL, sourceApplication: String?, annotation: AnyObject) -> Bool {
+        if SPTAuth.defaultInstance().canHandleURL(url) {
+            SPTAuth.defaultInstance().handleAuthCallbackWithTriggeredAuthURL(url, callback: { error, session in
+                if let error = error {
+                    // FIXME: throw application error
+                    print("openURL error: \(error.localizedDescription)")
+                    return
+                }
+                print("appDelegate recieved session: \(session)")
```
Comment (brightredchilli) [inline spotifyMini/AppDelegate.swift]: try not to leave these hanging print statements. the ones in error flows are usually an exception

File: spotifyMini/ViewController.swift [hunk 7]
```diff
@@ -1,25 +1,40 @@
...
-//  Copyright © 2016 intrepid. All rights reserved.
+//
 //
 
 import UIKit
 
-class ViewController: UIViewController {
-
+class ViewController: UIViewController, SPTAuthViewDelegate {
+
```
Comment (brightredchilli) [inline spotifyMini/ViewController.swift]: make sure your xcode setting will clean out trailing spaces on new lines.

File: spotifyMini/ViewController.swift [hunk 8]
```diff
@@ -1,25 +1,40 @@
...
+    }
+    
+    // MARK: SPTAuthViewDelegate
+    
+    func authenticationViewController(authenticationViewController: SPTAuthViewController!, didFailToLogin error: NSError!) {
+        print("login failed:\(error.localizedDescription)")
+    }
+    
+    func authenticationViewControllerDidCancelLogin(authenticationViewController: SPTAuthViewController!) {
+        print("login cancelled")
```
Comment (brightredchilli) [inline spotifyMini/ViewController.swift]: remove print statements

File: spotifyMini/ViewController.swift [hunk 9]
```diff
@@ -1,25 +1,40 @@
...
+    }
+    
+    func authenticationViewController(authenticationViewController: SPTAuthViewController!, didLoginWithSession session: SPTSession!) {
+        print("login success, session: \(session)")
+        // THIS fires if they don't have spotify,
```
Comment (brightredchilli) [inline spotifyMini/ViewController.swift]: good to turn this into a TODO that is actionable.

File: spotifyMini/ViewController.swift (hunk 7 above)
Comment (TOMalley104) [inline spotifyMini/ViewController.swift]: is that in Preferences > Text Editing > While Editing, check off "Automatically trim trailing whitespace" and "Include whitespace-only lines" ? Cause I checked that and it didn't fix this :(

Comment (brightredchilli) [inline spotifyMini/ViewController.swift]: its alright, just make sure you have them checked. I usually check my diff and make sure these don't creep in, it's more of a nit than a real style guide we follow at intrepid.

File: spotifyMini/ViewController.swift [hunk 10]
```diff
@@ -12,6 +12,8 @@ class ViewController: UIViewController, SPTAuthViewDelegate {
     
     override func viewDidLoad() {
         super.viewDidLoad()
+        NSNotificationCenter.defaultCenter().addObserver(self, selector: #selector(recieveSpotifyLogInNotification(_:)), name: SpotifyAppLogInSuccessNotification, object: nil)
```
Comment (brightredchilli) [inline spotifyMini/ViewController.swift]: make sure these get removed on dealloc

File: spotifyMini/AppDelegate.swift [hunk 11]
```diff
@@ -37,11 +38,10 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
             SPTAuth.defaultInstance().handleAuthCallbackWithTriggeredAuthURL(url, callback: { error, session in
                 if let error = error {
                     // FIXME: throw application error
-                    print("openURL error: \(error.localizedDescription)")
+                    NSNotificationCenter.defaultCenter().postNotificationName(SpotifyAppLogInSuccessNotification, object: nil, userInfo: ["error":error])
```
Comment (brightredchilli) [inline spotifyMini/AppDelegate.swift]: these keys usually get constantized too, but this is ok right now.

**Comment by brightredchilli** [comment]
:ship:

File: spotifyMini/ViewController.swift (hunk 10 above)
Comment (TOMalley104) [inline spotifyMini/ViewController.swift]: Good point! Forgot cause you don't have to do that as of iOS 9 [as of iOS 9](https://developer.apple.com/library/mac/releasenotes/Foundation/RN-Foundation/index.html#10_11NotificationCenter)

File: spotifyMini/AppDelegate.swift (hunk 11 above)
Comment (TOMalley104) [inline spotifyMini/AppDelegate.swift]: How do you like to do constants? would you recommend making a Constants class so it'd look more like `Constants.spotifyLogInSuccessNotification` ?

**Pull request merged**